### DISCUSSION TITLE
Calculating the percentage of cells expressing each gene

### DISCUSSION THREAD

**Discussion by Zig-E-Stardust** [satijalab/seurat#5033] [Q&A]
Hey people,
I wonder: is there a function/quick way to get a table where each row represents a gene and each column represents the percentage of cells with detectable expression of that gene in an individual sample (e.g. "mouse 1")?

Cheers,
Omer

**Comment by saketkc** [comment]
You can do this using `rowMeans`:
```r
rowMeans(GetAssayData(pbmc_small, assay="RNA", slot="counts"))
```

**Reply by Zig-E-Stardust** [reply]
Hey,
First, thanks a lot for the reply!
Second: I think this gives me the average expression of each gene. However, what I need is the percentage of cells expressing the gene (or, more accurately, the percentage of cells where expression of the gene was detected).

**Comment by samuel-marsh** [comment]
Hi,

There's couple ways you can do this:
- You can use this function: https://samuel-marsh.github.io/scCustomize/reference/Percent_Expressing.html
```
library(Seurat)
library(scCustomize)

percent <- PercentExpressing(seurat_object = OBJECT_NAME, features = c("GENE1", "GENE2", etc))
```
- You can calculate this manually by using `FetchData` and pulling the variables of interest and calculating percent of non-zero values per ident.  Or values above a certain threshold.  Note that `PercentExpressing` has extra parameter to specify a threshold too.
- You can pull from the ggplot2 object when using `DotPlot`.  ggplot2 objects contain both the visualization info and the data that underlies the plot.  Note that in this method for each ident the feature will have added numerical suffix but is still same gene but that can be confusing.
```
library(Seurat)
library(dplyr)
library(magrittr)

plot <- DotPlot(object = OBJECT_NAME, features = c("GENE1", "GENE2", etc))
plot_data <- plot$data %>% 
  select(pct.exp, id)
```

Best,
Sam

**Reply by Zig-E-Stardust** [reply]
Hey,
I've tried your first solution (using scCustomize) and it works like magic. That would be extremely useful for me - thanks a lot!
I would mention that you seem to have a tiny mistake - it should be "Percent_Expressing" and not "PercentExpressing".

Anyway, thanks a million!

**Comment by samuel-marsh** [comment]
Oh yep.  Definitely should be lol.  Funny to misspell name of function from my own package.

**Reply by Ruismart** [reply]
about the second way to only extract needed columns from Dotplot object, it's better to select another 'features.plot' column, because it's a long data type, the rownames are with additional '1/2/3/4/..' for a same feature
